// @progress/kendo-angular-gauges/dist/npm/base-components/scale.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var settings_component_1 = require("./settings.component");
/**
 * @hidden
 */
var ScaleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ScaleComponent, _super);
    function ScaleComponent() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    ScaleComponent.propDecorators = {
        labels: [{ type: core_1.Input }],
        majorTicks: [{ type: core_1.Input }],
        minorTicks: [{ type: core_1.Input }],
        min: [{ type: core_1.Input }],
        max: [{ type: core_1.Input }],
        minorUnit: [{ type: core_1.Input }],
        majorUnit: [{ type: core_1.Input }],
        reverse: [{ type: core_1.Input }],
        rangeSize: [{ type: core_1.Input }],
        rangePlaceholderColor: [{ type: core_1.Input }]
    };
    return ScaleComponent;
}(settings_component_1.SettingsComponent));
exports.ScaleComponent = ScaleComponent;
