// kinvey-angular-sdk/lib/index.js
Object.defineProperty(exports, "__esModule", { value: true });
var kinvey_html5_sdk_1 = require("kinvey-html5-sdk");
exports.getAppVersion = kinvey_html5_sdk_1.getAppVersion;
exports.setAppVersion = kinvey_html5_sdk_1.setAppVersion;
exports.logger = kinvey_html5_sdk_1.logger;
exports.Acl = kinvey_html5_sdk_1.Acl;
exports.Aggregation = kinvey_html5_sdk_1.Aggregation;
exports.StorageProvider = kinvey_html5_sdk_1.StorageProvider;
exports.DataStoreType = kinvey_html5_sdk_1.DataStoreType;
exports.Errors = kinvey_html5_sdk_1.Errors;
exports.Kmd = kinvey_html5_sdk_1.Kmd;
exports.Metadata = kinvey_html5_sdk_1.Kmd;
exports.Query = kinvey_html5_sdk_1.Query;
exports.AuthorizationGrant = kinvey_html5_sdk_1.AuthorizationGrant;
var datastore_service_1 = require("./datastore.service");
exports.DataStoreService = datastore_service_1.DataStoreService;
var endpoint_service_1 = require("./endpoint.service");
exports.EndpointService = endpoint_service_1.EndpointService;
var files_service_1 = require("./files.service");
exports.FilesService = files_service_1.FilesService;
var kinvey_module_1 = require("./kinvey.module");
exports.KinveyModule = kinvey_module_1.KinveyModule;
var ping_service_1 = require("./ping.service");
exports.PingService = ping_service_1.PingService;
var user_service_1 = require("./user.service");
exports.UserService = user_service_1.UserService;
