// @progress/kendo-angular-gauges/dist/npm/linear-gauge/pointer.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var services_1 = require("../services");
var base_components_1 = require("../base-components");
/**
 * The configuration options for a pointer item of a LinearGauge.
 */
var LinearPointerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LinearPointerComponent, _super);
    function LinearPointerComponent(configurationService, collectionChangesService) {
        return _super.call(this, configurationService, collectionChangesService) || this;
    }
    LinearPointerComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.ConfigurationService],
                    selector: 'kendo-lineargauge-pointer',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    LinearPointerComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    LinearPointerComponent.propDecorators = {
        border: [{ type: core_1.Input }],
        color: [{ type: core_1.Input }],
        margin: [{ type: core_1.Input }],
        opacity: [{ type: core_1.Input }],
        shape: [{ type: core_1.Input }],
        size: [{ type: core_1.Input }],
        value: [{ type: core_1.Input }]
    };
    return LinearPointerComponent;
}(base_components_1.CollectionItemComponent));
exports.LinearPointerComponent = LinearPointerComponent;
