// @progress/kendo-angular-gauges/dist/npm/linear-gauge/pointers.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var pointer_component_1 = require("./pointer.component");
var base_components_1 = require("../base-components");
var services_1 = require("../services");
/**
 * A collection of one or more LinearGauge pointers
 * ([more information]({% slug multiplepointers_lineargauge %})).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *     selector: 'my-app',
 *     template: `
 *         <kendo-lineargauge>
 *             <kendo-lineargauge-pointers>
 *                 <kendo-lineargauge-pointer *ngFor="let pointer of pointers"
 *                     [value]="pointer.value" [color]="pointer.color" shape="barIndicator">
 *                 </kendo-lineargauge-pointer>
 *             </kendo-lineargauge-pointers>
 *         </kendo-lineargauge>
 *     `
 * })
 * export class AppComponent {
 *     public pointers: any[] = [{
 *         value: 10,
 *         color: '#ff4500'
 *     }, {
 *         value: 12,
 *         color: '#28b4c8'
 *     }, {
 *         value: 20,
 *         color: '#8b0000'
 *     }];
 * }
 *
 * ```
 */
var LinearPointersComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LinearPointersComponent, _super);
    function LinearPointersComponent(configurationService, collectionChangesService) {
        return _super.call(this, 'pointer', configurationService, collectionChangesService) || this;
    }
    LinearPointersComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.CollectionChangesService],
                    selector: 'kendo-lineargauge-pointers',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    LinearPointersComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    LinearPointersComponent.propDecorators = {
        children: [{ type: core_1.ContentChildren, args: [pointer_component_1.LinearPointerComponent,] }]
    };
    return LinearPointersComponent;
}(base_components_1.CollectionComponent));
exports.LinearPointersComponent = LinearPointersComponent;
