// @progress/kendo-angular-gauges/dist/npm/linear-gauge/ranges.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var base_components_1 = require("../base-components");
var range_component_1 = require("./range.component");
var services_1 = require("../services");
/**
 * A collection of one or more LinearGauge scale ranges
 * ([more information and example]({% slug scaleranghes_lineargauge %})).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *     selector: 'my-app',
 *     template: `
 *         <kendo-lineargauge>
 *             <kendo-lineargauge-scale>
 *                 <kendo-lineargauge-scale-ranges>
 *                     <kendo-lineargauge-scale-range *ngFor="let range of ranges"
 *                         [from]="range.from" [to]="range.to" [color]="range.color">
 *                     </kendo-lineargauge-scale-range>
 *                 </kendo-lineargauge-scale-ranges>
 *             </kendo-lineargauge-scale>
 *         </kendo-lineargauge>
 *     `
 * })
 * export class AppComponent {
 *     public ranges: any[] = [{
 *         from: 0,
 *         to: 15,
 *         color: '#ffd246'
 *     }, {
 *         from: 15,
 *         to: 30,
 *         color: '#28b4c8'
 *     }, {
 *         from: 30,
 *         to: 50,
 *         color: '#78d237'
 *     }];
 * }
 *
 * ```
 */
var LinearRangesComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LinearRangesComponent, _super);
    function LinearRangesComponent(configurationService, collectionChangesService) {
        return _super.call(this, 'scale.ranges', configurationService, collectionChangesService) || this;
    }
    LinearRangesComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.CollectionChangesService],
                    selector: 'kendo-lineargauge-scale-ranges',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    LinearRangesComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    LinearRangesComponent.propDecorators = {
        children: [{ type: core_1.ContentChildren, args: [range_component_1.LinearRangeComponent,] }]
    };
    return LinearRangesComponent;
}(base_components_1.CollectionComponent));
exports.LinearRangesComponent = LinearRangesComponent;
