// @progress/kendo-angular-gauges/dist/npm/linear-gauge/scale.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var base_components_1 = require("../base-components");
var services_1 = require("../services");
/**
 * The configuration options for the scale of the LinearGauge
 * ([see example]({% slug scaleoptions_lineargauge %})).
 */
var LinearScaleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(LinearScaleComponent, _super);
    function LinearScaleComponent(configurationService) {
        var _this = _super.call(this, 'scale', configurationService) || this;
        _this.configurationService = configurationService;
        return _this;
    }
    LinearScaleComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-lineargauge-scale',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    LinearScaleComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService }
    ]; };
    LinearScaleComponent.propDecorators = {
        line: [{ type: core_1.Input }],
        ranges: [{ type: core_1.Input }],
        mirror: [{ type: core_1.Input }],
        vertical: [{ type: core_1.Input }]
    };
    return LinearScaleComponent;
}(base_components_1.ScaleComponent));
exports.LinearScaleComponent = LinearScaleComponent;
