// nativescript-picker/angular/nativescript-picker.directives.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var templated_items_comp_1 = require("nativescript-angular/directives/templated-items-comp");
var nativescript_picker_accessors_1 = require("./nativescript-picker.accessors");
var PickerFieldComponent = (function (_super) {
    __extends(PickerFieldComponent, _super);
    function PickerFieldComponent(_elementRef, _iterableDiffers) {
        return _super.call(this, _elementRef, _iterableDiffers) || this;
    }
    PickerFieldComponent_1 = PickerFieldComponent;
    Object.defineProperty(PickerFieldComponent.prototype, "nativeElement", {
        get: function () {
            return this.templatedItemsView;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(PickerFieldComponent.prototype, "className", {
        get: function () {
            return this._className;
        },
        enumerable: true,
        configurable: true
    });
    PickerFieldComponent.prototype.ngAfterContentInit = function () {
        _super.prototype.ngAfterContentInit.call(this);
        this.nativeElement.on("classNameChange", this.onClassNameChange.bind(this));
    };
    PickerFieldComponent.prototype.ngOnDestroy = function () {
        if (this.nativeElement) {
            this.nativeElement.off("classNameChange", this.onClassNameChange.bind(this));
        }
        _super.prototype.ngOnDestroy.call(this);
    };
    PickerFieldComponent.prototype.onClassNameChange = function (args) {
        this._className = args.object.className;
    };
    var PickerFieldComponent_1;
    PickerFieldComponent = PickerFieldComponent_1 = __decorate([
        core_1.Component({
            selector: "PickerField",
            template: "\n        <DetachedContainer>\n            <Placeholder #loader></Placeholder>\n        </DetachedContainer>",
            changeDetection: core_1.ChangeDetectionStrategy.OnPush,
            providers: [{ provide: templated_items_comp_1.TEMPLATED_ITEMS_COMPONENT, useExisting: core_1.forwardRef(function () { return PickerFieldComponent_1; }) }]
        }),
        __metadata("design:paramtypes", [core_1.ElementRef,
            core_1.IterableDiffers])
    ], PickerFieldComponent);
    return PickerFieldComponent;
}(templated_items_comp_1.TemplatedItemsComponent));
exports.PickerFieldComponent = PickerFieldComponent;
exports.DIRECTIVES = [PickerFieldComponent, nativescript_picker_accessors_1.PickerValueAccessor];
