// @progress/kendo-angular-gauges/dist/npm/arc-gauge/colors.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var base_components_1 = require("../base-components");
var color_component_1 = require("./color.component");
var services_1 = require("../services");
/**
 * A collection of one or more ArcGauge colors
 * ([more information and example]({% slug colorranges_arcgauge %})).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *     selector: 'my-app',
 *     template: `
 *         <kendo-arcgauge [value]="value">
 *              <kendo-arcgauge-colors>
 *                  <kendo-arcgauge-color *ngFor="let item of colors"
 *                      [from]="item.from" [to]="item.to" [color]="item.color">
 *                  </kendo-arcgauge-color>
 *              </kendo-arcgauge-colors>
 *         </kendo-arcgauge>
 *     `
 * })
 * export class AppComponent {
 *     public value: number = 10;
 *
 *     public colors: any[] = [{
 *         to: 25,
 *         color: '#0058e9'
 *     }, {
 *         from: 25,
 *         to: 50,
 *         color: '#37b400'
 *     }, {
 *         from: 50,
 *         to: 75,
 *         color: '#ffc000'
 *     }, {
 *         from: 75,
 *         color: '#f31700'
 *     }];
 * }
 *
 * ```
 */
var ColorsComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ColorsComponent, _super);
    function ColorsComponent(configurationService, collectionChangesService) {
        return _super.call(this, 'colors', configurationService, collectionChangesService) || this;
    }
    ColorsComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.CollectionChangesService],
                    selector: 'kendo-arcgauge-colors',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    ColorsComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    ColorsComponent.propDecorators = {
        children: [{ type: core_1.ContentChildren, args: [color_component_1.ColorComponent,] }]
    };
    return ColorsComponent;
}(base_components_1.CollectionComponent));
exports.ColorsComponent = ColorsComponent;
