// @progress/kendo-angular-gauges/dist/npm/arc-gauge/color.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var services_1 = require("../services");
var base_components_1 = require("../base-components");
/**
 * The configuration options for an ArcGauge color item.
 */
var ColorComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ColorComponent, _super);
    function ColorComponent(configurationService, collectionChangesService) {
        return _super.call(this, configurationService, collectionChangesService) || this;
    }
    ColorComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.ConfigurationService],
                    selector: 'kendo-arcgauge-color',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    ColorComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    ColorComponent.propDecorators = {
        color: [{ type: core_1.Input }],
        opacity: [{ type: core_1.Input }],
        from: [{ type: core_1.Input }],
        to: [{ type: core_1.Input }]
    };
    return ColorComponent;
}(base_components_1.CollectionItemComponent));
exports.ColorComponent = ColorComponent;
