// @progress/kendo-angular-gauges/dist/npm/arc-gauge/scale.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var base_components_1 = require("../base-components");
var services_1 = require("../services");
/**
 * The configuration options for the scale of the ArcGauge
 * ([see example]({% slug scaleoptions_arcgauge %})).
 */
var ArcScaleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(ArcScaleComponent, _super);
    function ArcScaleComponent(configurationService) {
        var _this = _super.call(this, 'scale', configurationService) || this;
        _this.configurationService = configurationService;
        return _this;
    }
    ArcScaleComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-arcgauge-scale',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    ArcScaleComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService }
    ]; };
    ArcScaleComponent.propDecorators = {
        labels: [{ type: core_1.Input }],
        rangeDistance: [{ type: core_1.Input }],
        rangeLineCap: [{ type: core_1.Input }],
        startAngle: [{ type: core_1.Input }],
        endAngle: [{ type: core_1.Input }]
    };
    return ArcScaleComponent;
}(base_components_1.ScaleComponent));
exports.ArcScaleComponent = ArcScaleComponent;
