// @progress/kendo-angular-label/dist/npm/label.directive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var kendo_angular_common_1 = require("@progress/kendo-angular-common");
/**
 * Represents the [Kendo UI Label directive for Angular]({% slug overview_label %}).
 * The Label associates a focusable Angular component or an HTML element
 * with a `label` tag by using the `[for]` property binding.
 *
 * To associate a component by using the `label` element, either:
 * * Set the `[for]` property binding to a
 * [template reference variable]({{ site.data.urls.angular['templatesyntax'] }}#template-reference-variables--var-), or
 * * Set the `[for]` property binding to an `id` HTML string value.
 *
 * @example
 * ```ts
 * _@Component({
 * selector: 'my-app',
 * template: `
 *  <div class="row example-wrapper" style="min-height: 450px;">
 *    <div class="col-xs-12 col-md-6 example-col">
 *      <label [for]="datepicker">DatePicker: </label>
 *      <kendo-datepicker #datepicker></kendo-datepicker>
 *    </div>
 *
 *    <div class="col-xs-12 col-md-6 example-col">
 *      <label [for]="'input'">Input: </label>
 *      <input id="input" />
 *    </div>
 *  </div>
 * `
 * })
 * class AppComponent { }
 * ```
 */
var LabelDirective = /** @class */ (function () {
    function LabelDirective() {
    }
    Object.defineProperty(LabelDirective.prototype, "labelFor", {
        get: function () {
            if (typeof this.for === 'string') {
                return this.for;
            }
            if (!kendo_angular_common_1.isDocumentAvailable()) {
                return null;
            }
            var component = this.getFocusableComponent() || {};
            return component.focusableId || component.id || null;
        },
        enumerable: true,
        configurable: true
    });
    /**
     * @hidden
     */
    LabelDirective.prototype.handleClick = function () {
        var component = this.getFocusableComponent();
        if (!component) {
            return;
        }
        if (component.focus) {
            component.focus();
        }
        // https://www.w3.org/TR/html52/sec-forms.html#labelable-element
        if (component instanceof HTMLInputElement && /^(checkbox|radio)$/.test(component.type) && !component.hidden) {
            component.click();
        }
    };
    LabelDirective.prototype.getFocusableComponent = function () {
        var target = this.for;
        return target && target.focus !== undefined ? target : null;
    };
    LabelDirective.decorators = [
        { type: core_1.Directive, args: [{
                    selector: 'label[for]' //tslint:disable-line:directive-selector
                },] },
    ];
    LabelDirective.propDecorators = {
        for: [{ type: core_1.Input }],
        labelFor: [{ type: core_1.HostBinding, args: ['attr.for',] }],
        handleClick: [{ type: core_1.HostListener, args: ['click',] }]
    };
    return LabelDirective;
}());
exports.LabelDirective = LabelDirective;
