// @progress/kendo-angular-gauges/dist/npm/radial-gauge/pointers.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var pointer_component_1 = require("./pointer.component");
var base_components_1 = require("../base-components");
var services_1 = require("../services");
/**
 * A collection of one or more RadialGauge pointers
 * ([more information and example]({% slug multiplepointers_radialgauge %})).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *     selector: 'my-app',
 *     template: `
 *         <kendo-radialgauge>
 *             <kendo-radialgauge-pointers>
 *                 <kendo-radialgauge-pointer *ngFor="let pointer of pointers"
 *                     [value]="pointer.value" [color]="pointer.color">
 *                 </kendo-radialgauge-pointer>
 *             </kendo-radialgauge-pointers>
 *         </kendo-radialgauge>
 *     `
 * })
 * export class AppComponent {
 *     public pointers: any[] = [{
 *         value: 10,
 *         color: '#ffd246'
 *     }, {
 *         value: 20,
 *         color: '#28b4c8'
 *     }, {
 *         value: 30,
 *         color: '#78d237'
 *     }];
 * }
 *
 * ```
 */
var RadialPointersComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RadialPointersComponent, _super);
    function RadialPointersComponent(configurationService, collectionChangesService) {
        return _super.call(this, 'pointer', configurationService, collectionChangesService) || this;
    }
    RadialPointersComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.CollectionChangesService],
                    selector: 'kendo-radialgauge-pointers',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    RadialPointersComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    RadialPointersComponent.propDecorators = {
        children: [{ type: core_1.ContentChildren, args: [pointer_component_1.RadialPointerComponent,] }]
    };
    return RadialPointersComponent;
}(base_components_1.CollectionComponent));
exports.RadialPointersComponent = RadialPointersComponent;
