// @progress/kendo-angular-gauges/dist/npm/radial-gauge/pointer.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var services_1 = require("../services");
var base_components_1 = require("../base-components");
/**
 * The configuration options for a pointer item of a RadialGauge.
 */
var RadialPointerComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RadialPointerComponent, _super);
    function RadialPointerComponent(configurationService, collectionChangesService) {
        return _super.call(this, configurationService, collectionChangesService) || this;
    }
    RadialPointerComponent.decorators = [
        { type: core_1.Component, args: [{
                    providers: [services_1.ConfigurationService],
                    selector: 'kendo-radialgauge-pointer',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    RadialPointerComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService },
        { type: services_1.CollectionChangesService }
    ]; };
    RadialPointerComponent.propDecorators = {
        cap: [{ type: core_1.Input }],
        color: [{ type: core_1.Input }],
        length: [{ type: core_1.Input }],
        value: [{ type: core_1.Input }]
    };
    return RadialPointerComponent;
}(base_components_1.CollectionItemComponent));
exports.RadialPointerComponent = RadialPointerComponent;
