// @progress/kendo-angular-gauges/dist/npm/radial-gauge/scale.component.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var tslib_1 = require("tslib");
var core_1 = require("@angular/core");
var base_components_1 = require("../base-components");
var services_1 = require("../services");
/**
 * The configuration options for the scale of the RadialGauge
 * ([more information and example]({% slug scaleoptions_radialgauge %})).
 */
var RadialScaleComponent = /** @class */ (function (_super) {
    tslib_1.__extends(RadialScaleComponent, _super);
    function RadialScaleComponent(configurationService) {
        var _this = _super.call(this, 'scale', configurationService) || this;
        _this.configurationService = configurationService;
        return _this;
    }
    RadialScaleComponent.decorators = [
        { type: core_1.Component, args: [{
                    selector: 'kendo-radialgauge-scale',
                    template: ''
                },] },
    ];
    /** @nocollapse */
    RadialScaleComponent.ctorParameters = function () { return [
        { type: services_1.ConfigurationService }
    ]; };
    RadialScaleComponent.propDecorators = {
        labels: [{ type: core_1.Input }],
        rangeDistance: [{ type: core_1.Input }],
        ranges: [{ type: core_1.Input }],
        startAngle: [{ type: core_1.Input }],
        endAngle: [{ type: core_1.Input }]
    };
    return RadialScaleComponent;
}(base_components_1.ScaleComponent));
exports.RadialScaleComponent = RadialScaleComponent;
