// @progress/kendo-angular-gauges/dist/es2015/radial-gauge/ranges.component.js
import { Component, ContentChildren, QueryList } from '@angular/core';
import { CollectionComponent } from '../base-components';
import { RadialRangeComponent } from './range.component';
import { CollectionChangesService, ConfigurationService } from '../services';
/**
 * A collection of one or more RadialGauge scale ranges
 * ([more information and example]({% slug scaleranghes_radialgauge %})).
 *
 * @example
 * ```ts
 * import { Component } from '@angular/core';
 *
 * _@Component({
 *     selector: 'my-app',
 *     template: `
 *         <kendo-radialgauge>
 *             <kendo-radialgauge-scale>
 *                 <kendo-radialgauge-scale-ranges>
 *                     <kendo-radialgauge-scale-range *ngFor="let range of ranges"
 *                         [from]="range.from" [to]="range.to" [color]="range.color">
 *                     </kendo-radialgauge-scale-range>
 *                 </kendo-radialgauge-scale-ranges>
 *             </kendo-radialgauge-scale>
 *         </kendo-radialgauge>
 *     `
 * })
 * export class AppComponent {
 *     public ranges: any[] = [{
 *         from: 0,
 *         to: 15,
 *         color: '#ffd246'
 *     }, {
 *         from: 15,
 *         to: 30,
 *         color: '#28b4c8'
 *     }, {
 *         from: 30,
 *         to: 50,
 *         color: '#78d237'
 *     }];
 * }
 *
 * ```
 */
export class RadialRangesComponent extends CollectionComponent {
    constructor(configurationService, collectionChangesService) {
        super('scale.ranges', configurationService, collectionChangesService);
    }
}
RadialRangesComponent.decorators = [
    { type: Component, args: [{
                providers: [CollectionChangesService],
                selector: 'kendo-radialgauge-scale-ranges',
                template: ''
            },] },
];
/** @nocollapse */
RadialRangesComponent.ctorParameters = () => [
    { type: ConfigurationService },
    { type: CollectionChangesService }
];
RadialRangesComponent.propDecorators = {
    children: [{ type: ContentChildren, args: [RadialRangeComponent,] }]
};
